File: R2_T03_05_Red_Kruzna_lista_Kolo_sesir.cpp
#include <iostream>
#include <list>
#include <string>
using namespace std;
int main()
{
    string s;
    cin >> s;                      // ULAZ: uuussuusis
    list<int> a;                   // ekvivalent LinkedList<int>
    for (int i = 0; i < s.length(); i++)
    {
        char c = s[i];
        if (c == 'u') a.push_front(a.size() + 1); // dodavanje novog igraca na pocetak
        else if (c == 'i' && !a.empty()) a.pop_front();// izbacivanje prvog
        else if (c == 's' && !a.empty())   // rotacija (poslednji ide na pocetak)
        {
            int ispred_v = a.back();  // poslednji
            a.pop_back();
            a.push_front(ispred_v);
        }
    }
    if (!a.empty())
    {
        int v = a.front();
        a.pop_front();
        a.push_back(v);
    }
    while (!a.empty())
    {
        cout << a.back() << " ";
        a.pop_back();
    }
    return 0;
}

File: R2_T03_05_Red_Kruzna_lista_Kolo_sesir.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String s = sc.next();

        Deque<Integer> a = new ArrayDeque<>();

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);

            if (c == 'u') {                     // dodavanje na pocetak
                a.addFirst(a.size() + 1);
            }
            else if (c == 'i' && !a.isEmpty()) { // izbacivanje prvog
                a.removeFirst();
            }
            else if (c == 's' && !a.isEmpty()) { // poslednji ide na pocetak
                int ispred_v = a.removeLast();
                a.addFirst(ispred_v);
            }
        }

        // priprema za stampu (prvi ide na kraj)
        if (!a.isEmpty()) {
            int v = a.removeFirst();
            a.addLast(v);
        }

        // stampanje (kao u originalu)
        while (!a.isEmpty()) {
            System.out.print(a.removeLast() + " ");
        }

        sc.close();
    }
}